src/libstd/panicking.rs

```diff
@@ -322,19 +324,36 @@ pub unsafe fn r#try<R, F: FnOnce() -> R>(f: F) -> Result<R, Box<dyn Any + Send>>
     let mut data = Data { f: ManuallyDrop::new(f) };
 
     let data_ptr = &mut data as *mut _ as *mut u8;
-    return if intrinsics::r#try(do_call::<F, R>, data_ptr, do_catch::<F, R>) == 0 {
-        Ok(ManuallyDrop::into_inner(data.r))
-    } else {
-        Err(ManuallyDrop::into_inner(data.p))
-    };
+    // SAFETY:
+    //
+    // Access to the union's fields: they are made safe by:
+    //  - 1) The call to `into_inner` will allow the data to be dropped later,
+    //    avoiding leaks.
+    //  - 2) The access to the field themselves is made safe by the check with
+    //    `intrinsics::r#try`.
+    //
+    // The call to `intrinsics::r#try` is made safe by:
+    //  - 1) A union is as large as its largest field.
+    //  - 2) See the comment above for the rest of the detailed explanation.
```

Comment:
These also seem somewhat like they're grasping at something to put rather than being truly helpful, I think we should go hunt down what the try intrinsic's arguments are and figure out appropriate preconditions. In this case I believe those are:
 * do_call, the first argument, can be called with the initial `data_ptr`.
 * do_catch, the second argument, can be called with the `data_ptr` as well.

The do_call and do_catch functions below should also be made unsafe and annotated with the preconditions their bodies imply.